added sample_weight support to K-Means and Minibatch K-means

xbhsu opened the pull request:
This addresses issue #3998 raised by @jnothman and adds sample_weight support to two more clustering algorithms. 

Centeroids in the K-Means algorithm are now given by the center of mass coordinates of the (weighted) samples and the algorithm now optimizes the weighted inertia. 

```
\sum_{i=0}^{n_clusters}\min_{\mu_j \in C} w_j (||x_j - \mu_i||^2)
```

where the centers are given by the center of mass of the clusters

```
\mu_j = \sum_{x_i \in C_j }^{n} w_i x_i / \sum_{x_i \in C_j} w_i
```

The signature of fit methods have been modified to `fit(self, X, y=None, sample_weight=None)` to be consistent elsewhere. sample_weight should be a numpy array of length `n_samples` containing the weight assignment of each sample. If sample_weight is None, the default, trivial weight of 1 is assigned to all the samples.

 In addition, tests were added to:
1. Check that the weighted inertia scales as expected with constant, non-trivial (e.g not equal to 1) weights
2. Check that constant, non-trivial weights yield identical label and center coordinates

Commit:
added weighted kmeans/minibatch

added c source

Commit:
added tests for weighted kmeans/minibatch

fixed typos

fix compatability issues

backward compatible np.newaxis

pep8 compliance

xbhsu commented:
I had to change
```python
np.multiply( sample_weight[center_mask, np.newaxis], X[center_mask])
```
to 
```python
np.multiply( sample_weight[center_mask][:, np.newaxis], X[center_mask])
```
Since the former is support in ```numpy >= 1.8.0``` and was causing Travis issues and failing tests using earlier versions of ```numpy```

agramfort commented:
does this change speed of non weighted code?

please run pep8 checker on your files. I saw some long lines

Commit:
inlined add csr to include multp. by weight

stop checking sample_weights if all ones

simplified sample weight checks

pep8

xbhsu commented on sklearn/cluster/_k_means.pyx:
```diff
@@ -349,10 +364,12 @@ def _centers_sparse(X, np.ndarray[INT, ndim=1] labels, n_clusters,
 
     for i in range(labels.shape[0]):
-        add_row_csr(data, indices, indptr, i, centers[labels[i]])
+        for ind in range(indptr[i], indptr[i + 1]):
```
implemented ```data[i,:] * sample_weight[i]``` directly in cython to improve performance on sparse ```X```

xbhsu commented:
Thanks for the feedback @agramfort! To get an idea of how the weighted code compares to the unweighted code speedy wise, I ran the ```fit``` method for ```KMeans``` and ```MinibatchKMeans``` with the weighted and unweighted code, averaging over ten runs. I've shared the gist [here](https://gist.github.com/76a89bb893c6d41ea2d2). For me, I got the following performance characteristics on dense and sparse inputs (labelled *_csr below) 
![figure_1](https://cloud.githubusercontent.com/assets/7433754/6121345/2eb07f7e-b0af-11e4-95ec-65157cdf5289.png)
The weighted code is a little slower, and the difference does grow with sample size. Perhaps its an issue with creating ```np.ones(shape=(n_samples,))``` or the calls to ```np.multiply```? Something that struck me as odd was that the sparse K-Means method (graph "kmeans_csr) is consistently much slower than K-means on dense inputs. I have yet to track down the source of this, but maybe you know the answer to that already?

agramfort commented:
it'a bit of a shame to reduce performance for the 98% usecase by adding
this feature.

GaelVaroquaux commented:
> it'a bit of a shame to reduce performance for the 98% usecase by adding
> this feature.

I must admit that I don't fully understand the usecase of a weighted
kmeans. Therefore I tend to agree with you.

amueller commented:
If the data is dense, sparse matrix formats are slower then dense formats, so your observation is expected.

jnothman commented:
> I must admit that I don't fully understand the usecase of a weighted
kmeans.

It's very possible I'm mistaken, and hoped for feedback at #3998. The
assumption is that sometimes you want to reduce your dataset to a set of
representative instances, but not all representatives indicate equal
density in the original space. The simplest example is where you have
duplicate objects (for instance I've been working with clustering documents
with similar sets of XPaths, and for some such signatures there are many
duplicates) and can't afford to represent each of the duplicated instances
separately in terms of time efficiency. BIRCH in theory should be
performing global clustering with its subcluster centres weighted, but atm
does not.

On 11 February 2015 at 04:43, Andreas Mueller <notifications@github.com>
wrote:

>  If the data is dense, sparse matrix formats are slower then dense
> formats, so your observation is expected.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/4218#issuecomment-73745901>
> .
>

xbhsu commented:
My understanding is that they're also used in modeling populations where the samples are not necessarily representative of the entire population (e.g. the proportion of subpopulations might be off). For instance, CDC, WHO, consumer spending etc. data often come with weights. People also use weights if they've collected data on a sample of consumers but are trying to extrapolate for a larger market.

agramfort commented:
how about a convincing example along these lines? I am bit scared of "this
could be useful in the near future".

amueller commented:
This might be useful here: #4357. I like the addition but I'd prefer it if it didn't have a performance impact if no weights are present. could you not make all the calculations (or at least the ones making it slower) conditional?

Commit:
speed improvements

speed improvements

added tests

pep8 compliance

stylistic changes

xbhsu commented:
Sorry for the delay. I got busy and then I broke something :)

Anyway, I've rewritten it so that the speed doesn't drop so dramatically. The weighting is now implemented conditionally.

I've also updated the [gist](https://gist.github.com/76a89bb893c6d41ea2d2.git) for the benchmark. I increased the n_init for miniBatchKMeans for a more stable benchmark. With this rewrite, the implementation performs comparably with the unweighted version. 
![figure_1](https://cloud.githubusercontent.com/assets/7433754/6623663/b1415fe4-c8ba-11e4-8aa1-82699c3c5e5d.png)

amueller commented:
I'd much prefer this, but I think your branching is not optimal yet. You have a lot of code duplication. Why don't you put the if around the places where there the code is actually different?

xbhsu commented:
I looked into that earlier, but when it's deep in the nested for-loops the ``if``` gets called a lot. That gave a noticeable slow down, especially for miniBatchKMeans.

amueller commented:
Huh, ok. I would have thought the if doesn't matter compared to the matrix operations, but I see now that these are actually vector-matrix operations. You do benchmark on pretty small data, though.
Too bad cython doesn't have real templates ^^

vote539 commented:
One example of an application for K-Means with sample weights is in the offline clustering step in the [Clustream](http://charuaggarwal.net/clstm.pdf) online clustering algorithm.  I wrote a Python implementation of Clustream (which I may submit as a PR of its own in a few weeks), but I only have a crude temporary solution for performing the offline clustering step until SKLearn's K-Means supports sample weights properly.

qinhanmin2014 commented:
Resolved in #10933, thanks @xbhsu
